### PR TITLE
Fixes for BP5 engine

### PR DESCRIPTION
**Pull request by franzpoeschel**
1. See [this comment](https://github.com/ornladios/ADIOS2/issues/2649#issuecomment-1058095101), the Span-based API is still very much relevant for performance in BP5.
2. The BP5 engine requires that every interaction with it happens inside an active IO step, take care of this

### CODE REVIEW

File: src/IO/ADIOS/ADIOS2IOHandler.cpp [hunk 1]
```diff
@@ -2189,7 +2200,8 @@ namespace detail
             {
                 isStreaming = true;
                 optimizeAttributesStreaming =
```
Comment (franzpoeschel) [inline src/IO/ADIOS/ADIOS2IOHandler.cpp:2210]: We can think about removing this optimization, it becomes irrelevant in variable-based iteration encoding and there is no harm in using that encoding for streaming.
Not relevant for this PR though.

File: src/IO/ADIOS/ADIOS2IOHandler.cpp [hunk 2]
```diff
@@ -2511,9 +2522,22 @@ namespace detail
     adios2::Engine &BufferedActions::requireActiveStep()
     {
         adios2::Engine &eng = getEngine();
+        /*
+         * If streamStatus is Parsing, do NOT open the step.
```
Comment (franzpoeschel) [inline src/IO/ADIOS/ADIOS2IOHandler.cpp:2540]: Should have added this comment earlier already, just took the opportunity now

File: src/IO/ADIOS/ADIOS2IOHandler.cpp [hunk 3]
```diff
@@ -2715,28 +2744,31 @@ namespace detail
...
-            adios2::StepStatus adiosStatus = m_lastStepStatus;
+            adios2::StepStatus adiosStatus{};
 
-            // Step might have been opened implicitly already
-            // by requireActiveStep()
-            // In that case, streamStatus is DuringStep and Adios
-            // return status is stored in m_lastStepStatus
             if (streamStatus != StreamStatus::DuringStep)
             {
-                flush(
```
Comment (franzpoeschel) [inline src/IO/ADIOS/ADIOS2IOHandler.cpp:2726]: Flushing here was wrong but harmless before. This dormant bug has woken up with this PR, so change it.

File: test/SerialIOTest.cpp [hunk 4]
```diff
@@ -5145,6 +5145,26 @@ TEST_CASE("iterate_nonstreaming_series", "[serial][adios2]")
                 backend.extension,
             false,
             backend.jsonBaseConfig());
+#if openPMD_HAVE_ADIOS2 &&                                                     \
+    ADIOS2_VERSION_MAJOR * 1000000000 + ADIOS2_VERSION_MINOR * 100000000 +     \
+            ADIOS2_VERSION_PATCH * 1000000 + ADIOS2_VERSION_TWEAK >=           \
+        2701001223
```
Comment (franzpoeschel) [inline test/SerialIOTest.cpp:5151]: Note that our CI does not yet have any ADIOS2 versions that satisfy these conditions

**Comment by franzpoeschel** [comment]
Reminder to myself: ReadRandomAccess

**Review by ax3l — APPROVED** [review]
Thank you!

Did I see correctly that we should backport this to `0.14.5`?

**Comment by ax3l** [comment]
As discussed: targeting BP5 patches for the 0.15.0 release only (no backports to 0.14.*)

**Pull request merged**